### PR TITLE
Sitting on chair near table gives bonuses to eating

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
`SUMMARY: Content "Sitting on chair near table gives mood bonus for eating, new trait for increased bonuses/penalties."`

#### Purpose of change
Tables and chairs are currently not so usesful, and mainly get smashed for materials. Now player has a reason to craft furniture, carry camp tables/chairs, mats, etc.

#### Describe the solution
When player stands on something sittable and there's a table (FLAT_SURF) nearby, player will get small mood bonus for eating.

Added new json flag "CAN_SIT" to everything sittable (chairs, armchairs, fiber mats, etc).
Added new trait "Table manners", which gives player bigger mood bonus for eating with table, but also gives mood penalty if player eats without table.

#### Describe alternatives you've considered
Not adding new purpose for tables and chairs, leaving them boring.
#### Additional context
Now fiber mats and leather tarps can really be used for picnics.

### CODE REVIEW

**Comment by Photoloss** [comment]
> gives mood penalty if player eats without table.

"Ate without table -20 morale" didn't know the Cataclysm had struck the Rimworld as well, RIP

**Comment by Night-Pryanik** [comment]
> When player stands on something sittable and there's a table nearby, player will get bonus to their crafting speed (10%).

Some things are better crafted standing still, not sitting. I suggest to require only table and/or decrease bonus to 5%.

**Comment by Mecares** [comment]
Water will boil faster if I sit down? Never noticed that in real life.

**Comment by Night-Pryanik** [comment]
Then it may be better to add new flag like "NO_BENEFIT_FROM_SITTING" and apply it to appropriate recipes such as boiling water.

**Comment by Inglonias** [comment]
I don't think that this should apply to crafting right now. Cooking in general is usually done standing. A bonus to mood when eating would be neat though.

**Comment by umsretuza** [comment]
Make chairs and tables more practically useful? Sounds like a great idea. However, there needs to be blacklists!

**Comment by kevingranade** [comment]
You're doing a few too many things at once here, please focus on either the
eating or crafting changes. Both of those are major balance changes and
need to be considered seperately.

**Comment by Wokko1** [comment]
Decreased bonuses and penalties, but made them longer.
Removed crafting bonus completely.
Renamed "Snobbish" to "Rigid table manners".
Excluded junk food from effects.

File: doc/JSON_FLAGS.md [hunk 1]
```diff
@@ -59,6 +59,7 @@ List of known flags, used in both terrain.json and furniture.json
 - ```BARRICADABLE_WINDOW_CURTAINS```
 - ```BASHABLE``` Players + Monsters can bash this.
 - ```BUTCHER_EQ``` Butcher's equipment - required for full butchery of corpses.
+- ```CAN_SIT``` Furniture the player can sit on. Player sitting near furniture with the "FLAT_SURF" tag will get mood bonus for eating and bonus to crafting speed.
```
Comment (Night-Pryanik) [inline doc/JSON_FLAGS.md]:
```suggestion
- ```CAN_SIT``` Furniture the player can sit on. Player sitting near furniture with the "FLAT_SURF" tag will get mood bonus for eating.
```

File: src/consumption.cpp [hunk 2]
```diff
@@ -710,6 +713,33 @@ bool player::eat( item &food, bool force )
         } else {
             add_msg_player_or_npc( _( "You eat your %s." ), _( "<npcname> eats a %s." ),
                                    food.tname().c_str() );
+            if ( !spoiled && !food.has_flag( "ALLERGEN_JUNK" ) ) {
+                bool has_table_nearby = false;
+                for ( const tripoint &pt : g->m.points_in_radius( pos(), 2 ) ) {
```
Comment (Night-Pryanik) [inline src/consumption.cpp]:
```suggestion
                for ( const tripoint &pt : g->m.points_in_radius( pos(), 1 ) ) {
```

Sitting two tiles from the table?

File: src/consumption.cpp [hunk 3]
```diff
@@ -710,6 +713,33 @@ bool player::eat( item &food, bool force )
...
+                    if ( g->m.has_flag_furn( "FLAT_SURF", pt ) || g->m.has_flag( "FLAT_SURF", pt ) ||
+                        ( g->m.veh_at(pt) && g->m.veh_at(pt)->vehicle().has_part( "KITCHEN" ) ) ) {
+                        has_table_nearby = true;
+                    }
+                }
+                bool eat_with_table = false;
+                if ( g->m.has_flag_furn( "CAN_SIT", pos() ) && has_table_nearby )
```
Comment (Night-Pryanik) [inline src/consumption.cpp]:
```suggestion
                if( g->m.has_flag_furn( "CAN_SIT", pos() ) && has_table_nearby ) {
```

Please astyle your changes. Also, use curly brackets even for one-line statements.

File: src/consumption.cpp [hunk 4]
```diff
@@ -710,6 +713,33 @@ bool player::eat( item &food, bool force )
...
+                    eat_with_table = true;
+                if ( eat_with_table ) {
+                    if ( has_trait( trait_id( "TABLEMANNERS" ) ) ) {
+                        if ( has_morale( MORALE_ATE_WITHOUT_TABLE ) )
```
Comment (Night-Pryanik) [inline src/consumption.cpp]: Redundant.

File: src/consumption.cpp (hunk 2 above)
Comment (Wokko1) [reply src/consumption.cpp]: Yeah, forgot about that.

Comment (Wokko1) [reply src/consumption.cpp]: Oh, I just thought radius of 1 is just one tile below the player.
Thanks.

**Comment by Wokko1** [comment]
Astyled the best I could, as I don't have git installed. Please let me know if I missed something else.

**Comment by Night-Pryanik** [comment]
> I don't have git installed

Astyle could be installed separately of git (as a plugin for VS or Notepad++, or completely stand-alone).

**Comment by Vasyan2006** [comment]
Will trait "Table manners" apply morale bonuses or penalties for using correct or wrong  cutlery?
Only dirty plebs will eat [fake french accent on] *viande cuite à la perfection* [fake french accent off] using inappropriate salad fork.

**Comment by Wokko1** [comment]
Gorgon and travis for some reason wrote "errors", but the errors are related to some magazine slots testing. Idk what that means, is it okay?

**Comment by I-am-Erk** [comment]
jenkins rebuild

**Comment by Mecares** [comment]
May i suggest:
"You've been taught proper table manners from your early childhood on. Now you can't even think about eating without a table.  Eating without it frustrates you, but eating like a civilized Person gives you a bigger mood bonus."

**Comment by Wokko1** [comment]
Thx, edited the description.

**Comment by Wokko1** [comment]
Jenkins rebuild

File: src/morale_types.h [hunk 5]
```diff
@@ -43,6 +43,8 @@ class morale_type_data
 extern const morale_type MORALE_NULL;
 extern const morale_type MORALE_FOOD_GOOD;
 extern const morale_type MORALE_FOOD_HOT;
+extern const morale_type MORALE_ATE_WITH_TABLE;
```
Comment (kevingranade) [inline src/morale_types.h]: No need to add more of these.
```suggestion
```

File: src/morale_types.h [hunk 6]
```diff
@@ -43,6 +43,8 @@ class morale_type_data
...
+extern const morale_type MORALE_ATE_WITHOUT_TABLE;
```
Comment (kevingranade) [inline src/morale_types.h]:
```suggestion
```

File: data/json/mutations.json [hunk 7]
```diff
@@ -271,6 +271,14 @@
         "starting_trait" : true,
         "valid" : false,
         "cancels" : ["SLOWLEARNER"]
+    },{
+        "type" : "mutation",
+        "id" : "TABLEMANNERS",
+        "name" : "Rigid table manners",
+        "points" : -1,
+        "description" : "You've been taught proper table manners from your early childhood on. Now you can't even think about eating without a table. Eating without it frustrates you, but eating like a civilized person gives you a bigger mood bonus.",
```
Comment (Night-Pryanik) [inline data/json/mutations.json]:
```suggestion
        "description" : "You've been taught proper table manners from your early childhood on.  Now you can't even think about eating without a table.  Eating without it frustrates you, but eating like a civilized person gives you a bigger morale bonus.",
```

File: src/consumption.cpp [hunk 8]
```diff
@@ -710,6 +713,32 @@ bool player::eat( item &food, bool force )
...
+                bool has_table_nearby = false;
+                for( const tripoint &pt : g->m.points_in_radius( pos(), 1 ) ) {
+                    if( g->m.has_flag_furn( "FLAT_SURF", pt ) || g->m.has_flag( "FLAT_SURF", pt ) ||
+                        ( g->m.veh_at( pt ) && g->m.veh_at( pt )->vehicle().has_part( "KITCHEN" ) ) ) {
+                        has_table_nearby = true;
+                    }
+                }
+                if( g->m.has_flag_furn( "CAN_SIT", pos() ) && has_table_nearby ) {
+                    if( has_trait( trait_id( "TABLEMANNERS" ) ) ) {
+                        if( has_morale( MORALE_ATE_WITHOUT_TABLE ) ) {
```
Comment (Night-Pryanik) [inline src/consumption.cpp]: This check is redundant. The next line with `rem_morale( MORALE_ATE_WITHOUT_TABLE )` will remove morale bonus no matter if player has this morale or not.

File: src/consumption.cpp [hunk 9]
```diff
@@ -710,6 +713,32 @@ bool player::eat( item &food, bool force )
...
+                            rem_morale( MORALE_ATE_WITHOUT_TABLE );
+                        }
+                        add_morale( MORALE_ATE_WITH_TABLE, 3, 3, 3_hours, 2_hours, true );
+                    } else {
+                        add_morale( MORALE_ATE_WITH_TABLE, 1, 1, 3_hours, 2_hours, true );
+                    }
+                } else {
+                    if( has_trait( trait_id( "TABLEMANNERS" ) ) ) {
+                        if( has_morale( MORALE_ATE_WITH_TABLE ) ) {
```
Comment (Night-Pryanik) [inline src/consumption.cpp]: This check is redundant. The next line with `rem_morale( MORALE_ATE_WITH_TABLE )` will remove morale bonus no matter if player has this morale or not.

**Comment by Wokko1** [comment]
Thanks for the help, guys.

**Comment by Night-Pryanik** [comment]
The next logical step is to take spoons, forks and knives into account when eating.

**Comment by adamkad1** [comment]
yeah, use cutlery when

**Comment by Little-119** [comment]
MREs include (plastic?) cutlery, so that would give them a use.

**Pull request merged**